nuitka/SourceCodeCaching.py

```diff
@@ -0,0 +1,423 @@
+
+    cache_name = makeCacheName(module_name, source_code)
+    tree_filename = _getCacheFilename(cache_name, "pkl")
+    meta_filename = _getCacheFilename(cache_name, "json")
+
+    # Validate cache metadata using shared helper
+    metadata = _validateCacheMetadata(meta_filename, tree_filename, module_name)
+    if metadata is None:
+        _cache_misses += 1
+        return None
+
+    # Validate cache file security before unpickling
+    if not _validateCacheFileSecurity(tree_filename):
+        general.warning(
+            "Cache file security validation failed for '%s'. Skipping cache." % module_name
+        )
+        _cache_misses += 1
+        return None
+
+    # Load the pickled AST
+    try:
+        with open(tree_filename, "rb") as f:
+            ast_tree = pickle.load(f)
+
+        # Record cache hit
+        _cache_hits += 1
+        _cache_restored_modules.append(module_name.asString())
+
+        return ast_tree
+
+    except Exception as e:
+        general.warning(
+            "Failed to load cached tree for module '%s': %s" % (module_name, e)
+        )
+        _cache_misses += 1
+        return None
+
+
+def writeCachedAST(module_name, source_code, ast_tree):
+    """Write an AST to the cache.
+
+    Args:
+        module_name: ModuleName object
+        source_code: String containing the source code
+        ast_tree: The Python AST object to cache
+    """
+    cache_name = makeCacheName(module_name, source_code)
+    tree_filename = _getCacheFilename(cache_name, "pkl")
+    meta_filename = _getCacheFilename(cache_name, "json")
+
+    try:
+        # Create cache directory if it doesn't exist
+        cache_dir = os.path.dirname(tree_filename)
+        makePath(cache_dir)
+
+        # Serialize the AST to a temporary file, then atomically move it
+        fd, temp_tree_filename = tempfile.mkstemp(dir=cache_dir, suffix=".pkl.tmp")
+        try:
+            with os.fdopen(fd, "wb") as f:
+                pickle.dump(ast_tree, f, protocol=pickle.HIGHEST_PROTOCOL)
+            replaceFileAtomic(temp_tree_filename, tree_filename)
+        except:
+            # Clean up temp file if something went wrong
+            if os.path.exists(temp_tree_filename):
+                os.unlink(temp_tree_filename)
+            raise
+
+        # Write metadata to temporary file, then atomically move it
+        fd, temp_meta_filename = tempfile.mkstemp(dir=cache_dir, suffix=".json.tmp")
+        try:
+            # Close the fd; writeJsonToFilename will open it
+            os.close(fd)
+
+            metadata = {
+                "file_format_version": _cache_format_version,
+                "module_name": module_name.asString(),
+                "nuitka_version": version_string,
+                "python_version": sys.version,
+            }
+            writeJsonToFilename(filename=temp_meta_filename, contents=metadata)
+            replaceFileAtomic(temp_meta_filename, meta_filename)
```

Comment:
Incomplete Json files would be skipped, if the writing of a json files is interrupted, it is automatically invalid, no need to use a temp file at all.